testing/upload_validation.ts

```diff
@@ -227,6 +228,120 @@ export function validateSyncTableSchema(
+    return result;
+  }
+  const syncTableSchemasByName: Record<string, ObjectSchema<any, any>> = {};
+  for (const syncTable of pack.syncTables) {
+    syncTableSchemasByName[syncTable.name] = syncTable.schema;
+  }
+  for (const [index, syncTable] of pack.syncTables.entries()) {
+    let parent: string | undefined;
+    for (const param of syncTable.getter.parameters) {
+      if (!param.crawlStrategy) {
+        continue;
+      }
+      if (param.crawlStrategy.parentTable) {
+        const {tableName, propertyKey} = param.crawlStrategy.parentTable;
+
+        const tableSchema = syncTableSchemasByName[tableName];
+        if (!tableSchema) {
+          context?.addIssue({
+            code: z.ZodIssueCode.custom,
+            path: ['syncTables', index],
+            message: `Sync table ${syncTable.name} expects parent table ${tableName} to exist.`,
+          });
+          return undefined;
+        }
+
+        const property = tableSchema.properties[propertyKey];
+        if (!property) {
+          context?.addIssue({
+            code: z.ZodIssueCode.custom,
+            path: ['syncTables', index],
+            message: `Sync table ${syncTable.name} expects parent table ${tableName}'s schema to have the property ${propertyKey}.`,
+          });
+          return undefined;
+        }
+
+        // TODO(patrick): Validate the types match
+
+        // We only allow one parent per table.
+        if (parent && parent !== tableName) {
+          context?.addIssue({
+            code: z.ZodIssueCode.custom,
+            path: ['syncTables', index],
+            message: `Sync table ${syncTable.name} cannot reference multiple parent tables.`,
+          });
+          return undefined;
+        }
+
+        const childList = result[tableName] || [];
+        // This table may already be in the child list if it uses multiple params from the parent.
+        if (!childList.includes(syncTable.name)) {
+          childList.push(syncTable.name);
+        }
+        result[tableName] = childList;
+        parent = tableName;
+      }
+    }
+  }
+  // Verify that there's no cycle
+  if (_hasCycle(result)) {
+    context?.addIssue({
+      code: z.ZodIssueCode.custom,
+      path: ['syncTables'],
+      message: `Sync table parent hierarchy is cyclic`,
+    });
+    return undefined;
+  }
+  return result;
+}
+
+// Exported for tests
+/** @hidden */
+export function _hasCycle(tree: Record<string, string[]>): boolean {
+  function subtreeHasCycle(currentKey: string, children: string[], visited: Set<string>): boolean {
+    if (visited.has(currentKey)) {
+      return true;
+    }
+    visited.add(currentKey);
+    for (const child of children) {
+      const subtree = tree[child];
+      if (!subtree) {
+        return false;
+      }
+      if (subtreeHasCycle(child, subtree, visited)) {
+        return true;
+      }
+    }
+    visited.delete(currentKey);
+    return false;
+  }
+
+  for (const key of Object.keys(tree)) {
```

Comment:
If you're feeling cheeky this could maybe be shortened to something like `return subtreeHasCycle('__CODA_INTERNAL_ROOT__', Object.keys(tree), new Set<string>())`

Reply:
So cheeky! I love it.